### PR TITLE
[VP] Fix short stream changes

### PR DESCRIPTION
**Pull request by peak3d**
## Description
Thx to @FernetMenta who provide a proper fix for the sync issues if we have short chapters devided by DMX_SPECIALID_STREAMCHANGE.

## Motivation and Context
If multi chapter streams start with a short into (in my example 3 sec), DMX_SPECIALID_STREAMCHANGE arrives from demuxer before the first packet was decoded and therefore player started. This leads to AV sync issues and  depending on the device to long chapter crossings.

## How Has This Been Tested?
Play Mandalorian S1E1 with Disney addon on NVIDIA Shield TV
Play VTM Go "Loïc: Zot van Koken"

## Types of change
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by michaelarnauts** [comment]
This is also a fix for https://github.com/peak3d/inputstream.adaptive/issues/332 and https://github.com/add-ons/plugin.video.vtm.go/issues/139

The previous attempt to fix this was https://github.com/xbmc/xbmc/pull/16658, but that PR was reverted.

**Comment by mediaminister** [comment]
I tested this a bit and I can confirm this fixes AV sync issues with one small period followed by a longer period.

But this also introduces a new problem with seeking. I got `eof reading from demuxer` when seeking with the up arrow key with VTM GO add-on.
Here is a full debug log:
[seekdebug.log](https://github.com/xbmc/xbmc/files/4509009/seekdebug.log)

**Comment by peak3d** [comment]
Thx, will look later at itVon meinem Huawei-Telefon gesendet-------- Ursprüngliche Nachricht --------Von: mediaminister <notifications@github.com>Datum: Di., 21. Apr. 2020, 11:34An: xbmc/xbmc <xbmc@noreply.github.com>Cc: Markus Pfau <pfau@peak3d.de>, Author <author@noreply.github.com>Betreff: Re: [xbmc/xbmc] [VP] Postpone DMX_SPECIALID_STREAMCHANGE until player is started (#17713)
I tested this a bit and I can confirm this fixes AV sync issues with one small period followed by a longer period.
But this also introduces a new problem with seeking. I got eof reading from demuxer when seeking with the up arrow key with VTM GO add-on.
Here is a full debug log:
seekdebug.log

**Comment by peak3d** [comment]
Forced pushed a change @mediaminister, thx for testing!

**Comment by mediaminister** [comment]
Thanks, I can confirm seeking with the up arrow key works perfectly now.

But when the stream starts with two small periods, there is still an AV sync issue when the third (long) period starts. Here is a full debug log:
[twosmallchapters.log](https://github.com/xbmc/xbmc/files/4514146/twosmallchapters.log)

**Comment by peak3d** [comment]
Yes, I know, will try to fix this, too

**Comment by peak3d** [comment]
@mediaminister I added a condition -> pls. check again

**Comment by mediaminister** [comment]
I tested but this doesn't solve it.
Log: [notyetsolved.log](https://github.com/xbmc/xbmc/files/4517278/notyetsolved.log)

**Comment by peak3d** [comment]
I'll add some more logging, I'm in opinion that your update was not successful.
I started multiple of the series you watch, and they are all fine now

**Comment by mediaminister** [comment]
>  I started multiple of the series you watch, and they are all fine now.

Did you try with "Loïc: Zot van Koken", because there is an extra sponsor advertisement before the program starts.
For instance: 
```
xbmc.executebuiltin('PlayMedia(plugin://plugin.video.vtm.go/play/catalog/episodes/cd5d1db0-519f-40de-b07e-bceb716e4848)')
```
in autoexec.py

For me it isn't solved on Linux x64. Here is a new log with the extra logging:
[notyetsolved_withlog.log](https://github.com/xbmc/xbmc/files/4517721/notyetsolved_withlog.log)

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 1]
```diff
@@ -1020,14 +1020,38 @@ bool CVideoPlayer::ReadPacket(DemuxPacket*& packet, CDemuxStream*& stream)
...
     if (packet->iStreamId == DMX_SPECIALID_STREAMCHANGE)
     {
+      CLog::Log(LOGDEBUG, "DMX: VSync: %d, ASync: %d, VAVSync: %d, AAVSync: %d",
+                m_CurrentVideo.syncState, m_CurrentAudio.syncState,
+                m_CurrentVideo.avsync, m_CurrentAudio.avsync);
+
+      if (m_CurrentVideo.id >= 0 && m_CurrentVideo.inited && m_CurrentAudio.id >= 0 &&
+          m_CurrentAudio.inited
+        && ((m_CurrentVideo.syncState == IDVDStreamPlayer::SYNC_STARTING
+          ||m_CurrentAudio.syncState == IDVDStreamPlayer::SYNC_STARTING) ||
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:1043]: > Conventional operators have to be surrounded by one space on each side.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by peak3d** [comment]
@mediaminister next try, thx for testing

**Comment by mediaminister** [comment]
Thanks, it plays perfectly now, all av sync issues seems to be fixed! 

But, the `eof reading from demuxer` when seeking with the up arrow key is back.

I tested with
```
xbmc.executebuiltin('PlayMedia(plugin://plugin.video.vtm.go/play/catalog/episodes/dbbf9404-323a-42b9-ad8c-65d651e6f303)')
```
in autoexec.py

Debug log: 
[seekbugagain.log](https://github.com/xbmc/xbmc/files/4519138/seekbugagain.log)

**Comment by peak3d** [comment]
@mediaminister the stream should now continue after seek, but - assumeing that you start seeking while in a short chapter - continues at the beginning of the chapter seeked. This is something I'll fix in a second step in inputstream.adaptive.

**Comment by mediaminister** [comment]
Yes, the stream continues after seeking, but the AV sync bug when a third (long) period starts is reintroduced:
[avsyncissue.log](https://github.com/xbmc/xbmc/files/4534885/avsyncissue.log)

**Comment by peak3d** [comment]
@mediaminister latest commit should be working now.

**Comment by mediaminister** [comment]
Thanks a lot @peak3d, I can confirm all AV sync issues are gone and seeking works:
[everythingok.log](https://github.com/xbmc/xbmc/files/4539120/everythingok.log)

**Comment by michaelarnauts** [comment]
OMG. It finally happened! @peak3d and @mediaminister you guys rule!

**Comment by phunkyfish** [comment]
I don't think this is a hack. It does what @FernetMenta  suggested in the post below. Postpone any further reads (or stream changes) until the current one is complete.

https://github.com/xbmc/xbmc/pull/16658#issuecomment-537350264

**Comment by fritsch** [comment]
I don't think so. He said properly wait unti a condition gets valid. The code here just sleeps 50 ms - so the condition is the "50 ms has ended-condition"?

I don't stand in the way of this PR, but it must be clear, that magic sleep values are meant to break, if we wait on something special to happen, we should wait on that "special thing" with an appropriate timeout - just sleeping here and there a bit is not good and will bite us back.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 2]
```diff
@@ -1020,14 +1020,35 @@ bool CVideoPlayer::ReadPacket(DemuxPacket*& packet, CDemuxStream*& stream)
...
 
   if (packet)
   {
     // stream changed, update and open defaults
     if (packet->iStreamId == DMX_SPECIALID_STREAMCHANGE)
     {
+      if (m_CurrentVideo.id >= 0 && m_CurrentAudio.id >= 0 && m_VideoPlayerVideo &&
+          m_VideoPlayerAudio && m_VideoPlayerVideo->HasData() &&
+          m_VideoPlayerAudio->HasData() &&
+          ((m_CurrentVideo.syncState != IDVDStreamPlayer::SYNC_INSYNC ||
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:1039]: reason for two `((` instead of one `(`?

**Comment by FernetMenta** [comment]
@fritsch is right. This will creates issues in particular with PVR.

This change does not consider conditions that clearly define that an ongoing stream change has completed or has even occurred. You can easily construct a case where this change will result in a hang. Note that on live TV a streamchange event can occur at any time:

streamchange event
some data
next streamchange event
postpone

VPA won't signal started because it has not enough data.

**Comment by michaelarnauts** [comment]
The same packet will be checked again after 50ms, so it's not some magic number where hopefully everything will be better, it's to give the player some time to start.

**Comment by fritsch** [comment]
The package is not checked, that's not the issue here. The "stop condition" - the && || stuff is used to know when it should go on, e.g. when the stream change package can be discarded. This package is just used to make a 50 ms delayed polling while assuming read is called again.

If internal player state does not change INSYNC, you will stay there forever.

**Comment by peak3d** [comment]
Not really true @fritsch, if VPV OR VPA run dry, the packed is processed as usual (not postponed)

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 2 above)
Comment (garbear) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:1039]: Also, the format of this complicated logic block is how clang-tidy would present it, but a more human-based format would increase understanding. Here is what I would do:

```c++

// Check for AV IDs
if (m_CurrentVideo.id >= 0 && m_CurrentAudio.id)
{
  // Check for AV players
  if (m_VideoPlayerVideo && m_VideoPlayerVideo->HasData() &&
      m_VideoPlayerAudio && m_VideoPlayerAudio->HasData())
  {
    // Check AV sync state
    if (m_CurrentVideo.syncState != IDVDStreamPlayer::SYNC_INSYNC ||
        m_CurrentAudio.syncState != IDVDStreamPlayer::SYNC_INSYNC ||
        m_CurrentVideo.avsync != CCurrentStream::AV_SYNC_NONE ||
        m_CurrentAudio.avsync != CCurrentStream::AV_SYNC_NONE)
     {
        CLog::Log(LOGDEBUG, "...");
        ...
        return true;
      }
    }
  }

  m_SelectionStreams.Clear(STREAM_NONE, STREAM_SOURCE_DEMUX);
  ...
```

The above is more understandable to me. Sure, it increases nesting, but I'd argue that's an acceptable engineering tradeoff in this situation.

**Comment by fritsch** [comment]
> Not really true @fritsch, if VPV OR VPA run dry, the packed is processed as usual (not postponed)
> 
> Edit: m_VideoPlayerVideo->HasData() in the condition. I tried so many other ways, but nothing fits together, its really over complicated, the comments on values do not match the behaviour.
> My personal opinion is, that VP (V/A) has to be rewritten.

Just seen your edit now. I agree with you, that understanding what really goes on is a huge problem and that's why we have this discussion. I see that some fanboys already stopped discussion on technical level and continued over with the "thumbs up" game, which means: We don't care at all, but, but, but, it works ... you don't that's highly appreciated :-)

In fact I checked HasData() before and was not sure about it being reliable for the actual use-case we have here. It's just the message queue size for either audio and video. Using it as a stop condition might work to not hang forever, but it's not a signal that also satisfies your INSYNC check. Means, if that happens, your code behave the very same way as quitting this loop cause of being now INSYNC.

If that's not a problem - fine with me. Reason: I don't have the oversight on what else might happen now in that possibility.

In fact I see two possibilities:
a) @FernetMenta gives us a more detailled tipp
b) We cope with the outtage as without a) I don't see how we can proceed. Knowledge is simply not there anymore.

**Comment by FernetMenta** [comment]
@fritsch Above I gave you an example why this code fails for live tv because VPA can't fill demuxer queue. Not sure how I can help here as long as my comments are ignored.

**Comment by FernetMenta** [comment]
> yes, that was the example for HasData() starts to return false

HasData does not return false for the scenario I gave because player wants to fill up demuxer queue to a certain level

> For the PVR example VPA queue would run empty and the code block would exit here. Correct?

Wrong, see above.

**Comment by fritsch** [comment]
Okay, got it, thx. That means we are actually stuck in that loop forever - and on the other scenario I mentioned if we exited because HasData() would be false, it's a wrong condition to assume streamchange is finished.

**Comment by behunin** [comment]
> My personal opinion is, that VP (V/A) has to be rewritten.

I share this opinion but have no idea who maintains VP, so I am glad to see a member share my opinion. I would like to know who is over VP though, it seams to me that no one is. I have ideas but don't know the code base well enough to make a PR and don't know who to talk to.

The idea of getting Kodi in line to stream them ALL is something I have been looking forward to, I think it has become obvious that this rewrite is the solution that will achieve that goal.

In short I think that VP needs to be abstracted to just the player and then extended or implemented by a class that handles all the changes to/from the player, as well as creation/deletion of players (Picture-in-Picture). Then data from a stream change can be delegated to another player, muted\skipped, held in a temp buffer then piped to a VP when ready, etc...

Curious what you had in mind @peak3d .

**Comment by phunkyfish** [comment]
Can we keep the PR discussion on topic please. A rewrite of VP is out of scope in this PR.

**Comment by peak3d** [comment]
@FernetMenta it only looks like we "overread" your comments. In fact (this is at least true for my person) I don't understand what you mean. I have not yet heared about a "Demux Queue". Like in the past: please be so kind and point to code lines where things can go wrong. I don't see any risk in the implementation as long VPA / VPV processes all packages. There is obly one thing I can think about: VPV/A are waiting for a prio event and demux packets are not processed - is this what you are talking about?

**Comment by matthuisman** [comment]
Do you think the issue of fast forwarding across periods is fixed by this as well?

Lots of episodes have about 6x short chapters at the end for credits.

This is what a user said
"Sometimes I'll fast forward to the end of an episode (e.g. "Clone Wars") because there is 4 minutes of credits before the next episode plays. This is when I notice the issue the most as it often buffers then tries to play the last 30 seconds. Only to end up starting over from the first 30 seconds."

**Comment by mediaminister** [comment]
> Do you think the issue of fast forwarding across periods is fixed by this as well?

I tested with VTM GO add-on and I can confirm that fast forward works across small periods.

**Comment by matthuisman** [comment]
That's with or without this PR?

**Comment by mediaminister** [comment]
> That's with or without this PR?

Fast forwarding works with both, so I guess I can't reproduce this problem with VTM GO add-on.

**Comment by phunkyfish** [comment]
Would something like this work:

```cpp
  // read a data frame from stream.
  if (m_pDemuxer)
  {
    packet = m_pDemuxer->Read();
    m_postponedTimer.SetInfinite();
  }

  if (packet)
  {
    // stream changed, update and open defaults
    if (packet->iStreamId == DMX_SPECIALID_STREAMCHANGE)
    {
      while (!StreamChangeCompleted() && !m_postponedTimer.TimePast())
      {
        CLog::Log(LOGDEBUG, "%s - Postponing DMX_SPECIALID_STREAMCHANGE until players are started",
                  __FUNCTION__);
        if (!m_postponingStreamChange)
        {
          m_postponedTimer.Set(5000);
          m_postponingStreamChange = true;
        }

        CThread::Sleep(50);
      }
      m_postponingStreamChange = false;
      ...
```

Obviously `StreamChangeCompleted()` still needs to defined based on the original or some other condition.

**Comment by phunkyfish** [comment]
@FernetMenta @fritsch does what I proposed above look reasonable?

**Comment by FernetMenta** [comment]
> does what I proposed above look reasonable?

no, you must not lock VP thread in a loop like this. if it is locked like this, it can't change the state

**Comment by matthuisman** [comment]
@FernetMenta 
How much do I need to pay you for you to tell us how to do this correctly?
My wife says I give a good back massage... Tempting?

This PR fixes quite a few issues so be nice to get a correct fix merged in :)

**Comment by peak3d** [comment]
@fritsch this PR is now for several weeks in the milhouse builds (rbpi / x86 linux) and I have not found a single negative report in his forum thread. How should we proceed? I use this PR for long time, too, without any issues.

If you know how to reproduce refering to the comment of @FernetMenta, can you pls. reproduce or guide me / us to get it reproduced?

**Comment by phunkyfish** [comment]
I reverted this change in my local build as it does indeed break with PVR. I couldn't reproduce reliably (it was on the family box in the living room) but I did have several instances where I had to power down the device. Since the revert zero power cycles.

**Comment by FernetMenta** [comment]
> @FernetMenta
> How much do we need pay you for you to tell us how to do this correctly?
> My wife says I give a good back massage...?
> 
> This PR fixes quite a few issues so be nice to get a correct fix merged in (and backported) :)

The problem here is that folks are egoistic and only interested in their own use cases, not showing respect for other scenarios which are not their own  priorities (pvr in this case). Even users with no coding knowledge at all chime in and try to push their use cases regardless if it breaks existing stuff.
Instead of discussing the design and trying to find a solution that is suitable for all, people defend this change and push others to prove that it is wrong.

Nevertheless

> tell us how to do this correctly?

It's that simple. Ask for help and you'll get it. Try this change (I did some time ago) first:
https://pastebin.com/KpH7GcRk
It fixes dynamic program change for pvr and might also fix this issue here. Even if it did not fix the issue here, it should be applied before further investigation.

**Comment by peak3d** [comment]
For me @FernetMenta  fix works (Disney+), curious if it also works with vtm go

**Comment by mediaminister** [comment]
I can confirm this works for VTM GO: [shortshapter.log](https://github.com/xbmc/xbmc/files/4701615/shortshapter.log)

**Comment by FernetMenta** [comment]
If you don't apply the full patch I provided, dynamic stream changes on live TV won't get fixed.

**Comment by peak3d** [comment]
The first hunk is not in current master, it was rejected. But let me search again

**Comment by FernetMenta** [comment]
@peak3d sorry, you are right. this was some debug helper code that never made it to master.

**Comment by peak3d** [comment]
Jenkins build this please

**Comment by peak3d** [comment]
IOS build issue unrelated
